rust/processor/src/gap_detector.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    models::processor_status::ProcessorStatus,
+    processors::ProcessedVersions,
+    schema::processor_status,
+    utils::{
+        database::{execute_with_better_error, PgDbPool},
+        util::parse_timestamp,
+    },
+};
+use ahash::AHashMap;
+use diesel::{upsert::excluded, ExpressionMethods};
+
+// Number of batches processed before gap detected
+pub const GAP_DETECTION_BATCH_COUNT: u64 = 50;
+// Number of batches to process before updating processor status
+const PROCESSOR_STATUS_UPDATE_BATCH_COUNT: u64 = 10;
+
+pub struct GapDetector {
+    processor_name: String,
+    db_pool: PgDbPool,
+    starting_version: u64,
+    seen_versions: AHashMap<u64, ProcessedVersions>,
+    maybe_prev_batch: Option<ProcessedVersions>,
+    num_batches_processed_without_gap: u64,
+    num_batches_processed_with_gap: u64,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum GapDetectorResult {
+    GapDetected { gap_start_version: u64 },
+    NoGapDetected(),
+}
+
+impl GapDetector {
+    pub fn new(processor_name: String, db_pool: PgDbPool, starting_version: u64) -> Self {
+        Self {
+            processor_name,
+            db_pool,
+            starting_version,
+            seen_versions: AHashMap::new(),
+            maybe_prev_batch: None,
+            num_batches_processed_without_gap: 0,
+            num_batches_processed_with_gap: 0,
+        }
+    }
+
+    pub async fn process_versions(
```

Comment:
this no longer needs to be async :-)

Reply:
this needs to be async so that it can update processor status in DB. haven't added the write to file part yet but once that's built, we can make this non-async!